### PR TITLE
[SPARK-39704][SQL] Implement createIndex & dropIndex & indexExists in JDBC (H2 dialect)

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
Implementing createIndex/dropIndex/indexExists in DS V2 JDBC for H2 dialect.

### Why are the changes needed?
This is a subtask of the V2 Index support(https://issues.apache.org/jira/browse/SPARK-36525). 
This PR implements createIndex, dropIndex and indexExists. 
After review for some changes in this PR, I will create new PR for listIndexs.
**It can better test the index interface locally.**
> This PR only implements createIndex, IndexExists and dropIndex in H2 dialect.

### Does this PR introduce _any_ user-facing change?
Yes, createIndex/dropIndex/indexExists in DS V2 JDBC for H2

### How was this patch tested?
New UT.

### CODE REVIEW

**Comment by panbingkun** [comment]
ping @huaxingao

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala [hunk 1]
```diff
@@ -103,6 +106,40 @@ private[sql] object H2Dialect extends JdbcDialect {
     functionMap.clear()
   }
 
+  override def createIndex(
+    indexName: String,
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala:110]: nit: 4 space-indentation

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala [hunk 2]
```diff
@@ -103,6 +106,40 @@ private[sql] object H2Dialect extends JdbcDialect {
...
+    s"CREATE INDEX ${indexNameWithSchema(tableIdent, indexName)} $indexType ON " +
+      s"${tableNameWithSchema(tableIdent)} (${columnList.mkString(", ")})"
+  }
+
+  override def dropIndex(indexName: String, tableIdent: Identifier): String = {
+    s"DROP INDEX ${indexNameWithSchema(tableIdent, indexName)}"
+  }
+
+  override def indexExists(
+    conn: Connection,
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala:126]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala (hunk 1 above)
Comment (panbingkun) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala]: Updated

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala [hunk 3]
```diff
@@ -74,6 +77,40 @@ private[sql] object H2Dialect extends JdbcDialect {
     functionMap.clear()
   }
 
+  override def createIndex(
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala:80]: Could you include the link for H2 `CREATE INDEX` here?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala [hunk 4]
```diff
@@ -74,6 +77,40 @@ private[sql] object H2Dialect extends JdbcDialect {
...
+      indexName: String,
+      tableIdent: Identifier,
+      columns: Array[NamedReference],
+      columnsProperties: util.Map[NamedReference, util.Map[String, String]],
+      properties: util.Map[String, String]): String = {
+    val columnList = columns.map(col => col.fieldNames.head)
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala:86]: I think we always need to quote the identifier.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala [hunk 5]
```diff
@@ -74,6 +77,40 @@ private[sql] object H2Dialect extends JdbcDialect {
...
+    val (indexType, _) = JdbcUtils.processIndexProperties(properties, "h2")
+    s"CREATE INDEX ${indexNameWithSchema(tableIdent, indexName)} $indexType ON " +
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala:88]: Is the schema required here for index name? If not putting schema, does it default to the same schema as the table schema?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala [hunk 6]
```diff
@@ -74,6 +77,40 @@ private[sql] object H2Dialect extends JdbcDialect {
...
+      s"${tableNameWithSchema(tableIdent)} (${columnList.mkString(", ")})"
+  }
+
+  override def dropIndex(indexName: String, tableIdent: Identifier): String = {
+    s"DROP INDEX ${indexNameWithSchema(tableIdent, indexName)}"
+  }
+
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala:95]: Could you include the link for H2 SELECT INDEXES here?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala [hunk 7]
```diff
@@ -74,6 +77,40 @@ private[sql] object H2Dialect extends JdbcDialect {
...
+  override def indexExists(
+      conn: Connection,
+      indexName: String,
+      tableIdent: Identifier,
+      options: JDBCOptions): Boolean = {
+    val sql = s"SELECT * FROM INFORMATION_SCHEMA.INDEXES WHERE " +
+      s"TABLE_NAME = '${tableIdent.name()}' AND INDEX_NAME = '$indexName'"
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala:102]: Why the table name here doesn't need to include schema?

**Review by huaxingao — APPROVED** [review]

**Comment by huaxingao** [comment]
Merged to master. Thanks @panbingkun

**Pull request closed without merging**